### PR TITLE
feat: [CDS-46724]: Support for Tas Entities for NG

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes
1. Adding Infrastructure, Service, Connector for TAS in NG
2. Adding Instance sync for TAS in NG
3. Adding 3 new manifest types(TasManifest, Vars, Autoscaler)
4. Adding TAS connector validation NG task
5. Adding Instance sync for TAS containing TasInstanceSyncPerpetualTaskExecuter for delegate side.

Test Performed:
1. Creating TAS Service with TasManifest, Vars and AutoScaler manifest.
2. Created TAS connector.
3. Tested connector connectivity with harness platform
4. Tested connector connectivity with delegate

Screenshot 2022-11-18 at 10 40 39 AM

Screenshot 2022-11-18 at 10 41 19 AM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/40242)

### CODE REVIEW

**Review by anilchowdhury-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40242)* status: 124 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureValidator.java` line 252 at r9](https://reviewable.io/reviews/harness/harness-core/40242#-NH4IyPJEGMEcKJKzBC5:-NH4IyPJEGMEcKJKzBC6:bm5kvfb) ([raw file](https://github.com/harness/harness-core/blob/cfea29c9faeb89a4be7087ca22c033631bf8b462/125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureValidator.java#L252)):*
> ```Java
>     if (null == infrastructure.getOrganization()) {
>       throw new InvalidArgumentsException(Pair.of("Organization", "cannot be empty"));
>     }
> ```

should use hasValueOrExpression for organization

**Review by anilchowdhury-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40242)* status: 124 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureValidator.java` line 255 at r9](https://reviewable.io/reviews/harness/harness-core/40242#-NH4J41L8QfFfSKLAp2b:-NH4J41L8QfFfSKLAp2c:b-twt1su) ([raw file](https://github.com/harness/harness-core/blob/cfea29c9faeb89a4be7087ca22c033631bf8b462/125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureValidator.java#L255)):*
> ```Java
> 
>     if (null == infrastructure.getSpace()) {
>       throw new InvalidArgumentsException(Pair.of("Space", "cannot be empty"));
> ```

should use hasValueOrExpression for space

**Review by anilchowdhury-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40242)* status: 124 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`126-instance/src/main/java/io/harness/dtos/instanceinfo/InstanceInfoDTO.java` line 30 at r9](https://reviewable.io/reviews/harness/harness-core/40242#-NH4K3NKBTHRjTbBcMRN:-NH4K3NKBTHRjTbBcMRO:b53w1zh) ([raw file](https://github.com/harness/harness-core/blob/cfea29c9faeb89a4be7087ca22c033631bf8b462/126-instance/src/main/java/io/harness/dtos/instanceinfo/InstanceInfoDTO.java#L30)):*
> ```Java
>       @JsonSubTypes.Type(value = CustomDeploymentInstanceInfoDTO.class, name = "CustomDeployment"),
>       @JsonSubTypes.Type(value = EcsInstanceInfoDTO.class, name = "ECS"),
>       @JsonSubTypes.Type(value = TasInstanceInfoDTO.class, name = "Tas")
> ```

We should define this as "TAS"

**Review by anilchowdhury-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40242)* status: 124 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`440-connector-nextgen/src/main/java/io/harness/serializer/morphia/ConnectorMorphiaClassesRegistrar.java` line 137 at r9](https://reviewable.io/reviews/harness/harness-core/40242#-NH4MCOQC_Qwd-3hGvyf:-NH4MCOQC_Qwd-3hGvyg:blce9jp) ([raw file](https://github.com/harness/harness-core/blob/cfea29c9faeb89a4be7087ca22c033631bf8b462/440-connector-nextgen/src/main/java/io/harness/serializer/morphia/ConnectorMorphiaClassesRegistrar.java#L137)):*
> ```Java
>     set.add(AwsConfig.class);
>     set.add(SpotConfig.class);
>     set.add(TasConfig.class);
> ```

Why we are registering it twice

**Review by anilchowdhury-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40242)* status: 124 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`954-connector-beans/src/main/java/io/harness/delegate/beans/connector/tasconnector/TasConnectorDTO.java` line 44 at r9](https://reviewable.io/reviews/harness/harness-core/40242#-NH4Nycu17wy2dHEBo1M:-NH4Nycu17wy2dHEBo1N:b-lbxta8) ([raw file](https://github.com/harness/harness-core/blob/cfea29c9faeb89a4be7087ca22c033631bf8b462/954-connector-beans/src/main/java/io/harness/delegate/beans/connector/tasconnector/TasConnectorDTO.java#L44)):*
> ```Java
>     return null;
>   }
> ```

Should we define outcome for TAS Connector which can be used in Deployment Template

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/tasconnector/outcome/TasCredentialOutcomeDTO.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import lombok.experimental.FieldDefaults;
+
+@OwnedBy(HarnessTeam.CDP)
+@Data
+@NoArgsConstructor
+@JsonIgnoreProperties(ignoreUnknown = true)
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class TasCredentialOutcomeDTO {
+  @NotNull @JsonProperty("type") TasCredentialType type;
+  @JsonProperty("spec")
```
Comment (RishabhGupta34) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/tasconnector/outcome/TasCredentialOutcomeDTO.java:26]: We should remove the `JsonProperty` field from new outcome classes

File: 932-connector-task/src/main/java/io/harness/connector/task/tas/TasValidationHandler.java [hunk 2]
```diff
@@ -49,9 +49,16 @@ public ConnectorValidationResult validate(TasTaskParams tasTaskParams, List<Encr
   private ConnectorValidationResult validateInternal(
       TasConnectorDTO tasConnectorDTO, List<EncryptedDataDetail> encryptedDataDetails) {
     CloudFoundryConfig cfConfig = ngConfigMapper.mapTasConfigWithDecryption(tasConnectorDTO, encryptedDataDetails);
+    reformatEndpointURL(cfConfig);
     return handleValidateTask(cfConfig);
   }
 
+  private void reformatEndpointURL(CloudFoundryConfig cfConfig) {
+    String endpointUrl = cfConfig.getEndpointUrl();
+    if (endpointUrl.startsWith("https://"))
```
Comment (RishabhGupta34) [inline 932-connector-task/src/main/java/io/harness/connector/task/tas/TasValidationHandler.java:58]: this can be `http` as well right?

**Review by anilchowdhury-harness — APPROVED** [review]
:lgtm:

> Reviewed 3 of 71 files at r1, 16 of 24 files at r13, 13 of 13 files at r14, 79 of 86 files at r15, 4 of 6 files at r16, 12 of 12 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40242#-:-NHZFzMg9588s2towieU:bnfp4nl)* status: :shipit: complete! ✓

**Comment by sourabh-awashti** [comment]
trigger ti0

**Comment by RishabhGupta34** [comment]
trigger tin0

**Pull request closed without merging**